### ISSUE TITLE
Allow for batching stored procedure calls

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#6813]
Currently, we cannot batch stored procedure calls through the jOOQ API as is possible through JDBC

----
See also:
- https://groups.google.com/forum/#!topic/jooq-user/o0Zd9rKoxlg
- https://stackoverflow.com/q/47185044/521799

**Comment by viveleroi** [comment]
Any *real* plans to implement this?

**Comment by lukaseder** [comment]
@viveleroi What's the specific use-case in your case? Can you show some example code (e.g. involving JDBC) that shows what you'd like to achieve here?

**Comment by viveleroi** [comment]
Sure thing. I have a java application that logs activity data for Minecraft servers. Granted that's not a common use for tools like this so I'm not surprised it's not something most people need. It records events so that server owners can audit what's been happening - they can search and find rule violators, reverse changes to undo abuse, etc. Sometimes changes happen in bulk but we always queue them and insert in batches for perf reasons.

We now are adding support for stored procedures. The main procedure itself offloads some relational logic to the database server. It finds or creates entries in other tables and then INSERTS an activity record using those primary keys. Until now we've always run this logic from the app and cached those results. 

So I prepare a call:

`statement = connection.prepareCall("{ CALL createActivity(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) }")`

And iterate through the queue in a repeating task adding data to the statement and calling `statement.addBatch();`, and finally executing the batch as usual.

**Comment by lukaseder** [comment]
Thanks for the explanation. As a quick fix, it does seem to make sense to batch such calls. Batching (as in the JDBC sense) allows for reducing the number of server round trips by aggregating statements into a single statement.

But often, it's just a quick fix, reducing only the roundtrips. For example, with batching, you still don't profit from the optimisations an RDBMS can implement when bulk inserting data, for example. E.g. a bulk insert (an insert with multiple rows, e.g. `INSERT .. VALUES (...), (...), (...)`) can help the RDBMS better allocate disk resources because it knows exactly how many rows are going to be inserted. This isn't the case with batching.

In the case of stored procedures, it would be totally possible to create a procedure that accepts collections of data, e.g. `createActivities()`. At least, it's possible to do so in some RDBMS (you haven't stated which one it is). I've explored this in this article here for Oracle:
https://blog.jooq.org/the-cost-of-jdbc-server-roundtrips/

I'm sure it holds true for at least PostgreSQL and SQL Server as well, whose procedures can accept collections. In other RDBMS, I guess collections would have to be encoded as JSON or XML, which complicates things. But just to be sure that workarounds / alternative solutions aren't overlooked.

I'm not too excited about supporting batching for procedures in jOOQ via new API. It might work for `IN` parameter only procedures, but it seems very complicated for procedures producing `OUT` parameters or undeclared cursors. I'd have to explore this a bit, first. Adding new API is hard in this area, but I think if your procedure has `IN` parameters only, then at least the `BatchedConnection` should work:
https://www.jooq.org/doc/latest/manual/sql-execution/batched-connection/

It probably doesn't? In that case, that would be a bug. I'll investigate this, soon.

Another option that moves logic to the server would be to create anonymous blocks containing `CALL` statements:
- https://www.jooq.org/doc/latest/manual/sql-building/procedural-statements/procedural-block/
- https://www.jooq.org/doc/latest/manual/sql-building/procedural-statements/procedural-call/

There's still some convenience missing to turn a type safe, generated `Routine` call into a `CALL` statement: https://github.com/jOOQ/jOOQ/issues/11363, but it can be done manually, for sure.

**Comment by viveleroi** [comment]
Thanks I'll look into all of this and see what makes sense for my case.

My currently supported DBs are MySQL, MariaDB, and Postgres (and H2 but I have a non-procedure-based fallback). A quick search says mysql doesn't support collections as procedure inputs. My non-procedure fallback is also used for users who have mysql/etc but don't have permission to create routines - often the case if they're using a database provided by their server hosting company. In *those* cases I care more about server round trips so batching makes sense.

But when a user does have procedure permissions, they're more likely to be running their db server and I would assume is local - running on the same machine, but I don't have any stats to know if that's true.

I have no issue continuing to use JDBC, I understand this is a niche need.

Is there a way I could use jooq to generate *just* the CALL sql? At the very least I could reduce the plain sql in my code.

**Comment by lukaseder** [comment]
> I have no issue continuing to use JDBC, I understand this is a niche need.

Yeah, I guess that's the best way forward for now, though I do think that the `BatchedConnection` approach should work

> Is there a way I could use jooq to generate _just_ the CALL sql? At the very least I could reduce the plain sql in my code.

Well, the [`CALL` statement does that](https://www.jooq.org/doc/latest/manual/sql-building/procedural-statements/procedural-call/). You can always use an `ExecuteListener` to extract the generated SQL, and then abort execution with an exception, as a kind of hack...
